apps/communications/contacts/views/form/js/form_controller.js

```diff
@@ -0,0 +1,201 @@
+  }
+
+  function doMatch(contact, callbacks) {
+    LazyLoader.load(['/shared/js/text_normalizer.js',
+                     '/shared/js/simple_phone_matcher.js',
+                     '/shared/js/contacts/contacts_matcher.js'], function() {
+      Matcher.match(contact, 'active', callbacks);
+    });
+  }
+
+  var mergeHandler;
+  function cookMatchingCallbacks(contact) {
+    return {
+      onmatch: function(results) {
+        MatchService.showDuplicateContacts();
+
+        mergeHandler = function mergeHandler(e) {
+          if (e.origin !== CONTACTS_APP_ORIGIN) {
+            return;
+          }
+
+          var data = e.data;
+          switch (data.type) {
+            case 'duplicate_contacts_loaded':
+              // UI ready, passing duplicate contacts
+              var duplicateContacts = {};
+              Object.keys(results).forEach(function(id) {
+                duplicateContacts[id] = {
+                  matchingContactId: id,
+                  matchings: results[id].matchings
+                };
+              });
+
+              window.postMessage({
+                type: 'show_duplicate_contacts',
+                data: {
+                  name: getCompleteName(getDisplayName(contact)),
+                  duplicateContacts: duplicateContacts
+                }
+              }, CONTACTS_APP_ORIGIN);
+
+            break;
+
+            case 'merge_duplicate_contacts':
+              window.removeEventListener('message', mergeHandler);
+
+              // List of duplicate contacts to merge (identifiers)
+              var list = [];
+              var ids = data.data;
+              Object.keys(ids).forEach(id => {
+                list.push(results[id]);
+              });
+
+              doMerge(contact, list, function finished() {
+                window.postMessage({
+                  type: 'duplicate_contacts_merged',
+                  data: ids
+                }, CONTACTS_APP_ORIGIN);
+                close();
+
+              });
+
+            break;
+
+            case 'ready':
+              // TODO Use if needed
+              break;
+
+            case 'window_close':
+              // If user igonores duplicate contacts we save the contact
+              window.removeEventListener('message', mergeHandler);
+              doSave(contact, true);
+              break;
+          }
+        };
+
+        window.addEventListener('message', mergeHandler);
+      },
+      onmismatch: function() {
+        // Saving because there aren't duplicate contacts
+        doSave(contact);
+      }
+    };
+  }
+
+
+  function doSave(contact) {
+    ContactsService.save(contact, function(error) {
+      // TODO Use if needed.
+    });
+  }
+
+  exports.Form = {
+    init: function() {
+
+      _ = navigator.mozL10n.get;
+      window.addEventListener('close-ui', function() {
+        close();
+      });
+
+      window.addEventListener('save-contact', function(event) {
```

Comment:
nit:
Let's improve readability and extract this to a named function and do something like:
```
function saveContact() { ... }
window.addEventListener('save-contact', saveContact);